process_console: Added documentation (#7)
UPB-CS-OpenSourceUpstream/tock: doc/Process_Console.md

 pub struct Capability;
 unsafe impl capabilities::ProcessManagementCapability for Capability {}

 let pconsole = static_init!(
     ProcessConsole<usart::USART>,
     ProcessConsole::new(&usart::USART0,
                  115200,
                  &mut console::WRITE_BUF,
                  &mut console::READ_BUF,
                  &mut console::COMMAND_BUF,
                  kernel,
                  Capability));
 hil::uart::UART::set_client(&usart::USART0, pconsole);

 pconsole.start();
 ```

 Using ProcessConsole
 --------------------

 With this capsule properly added to a board's `main.rs` and that kernel
 loaded to the board, make sure there is a serial connection to the board.
 Likely, this just means connecting a USB cable from a computer to the board.
 Next, establish a serial console connection to the board. An easy way to do
 this is to run:

 ```shell
 $ tockloader listen
 ```

 With that console open, you can issue commands. For example, to see all of
 the processes on the board, use `list`:

 ```text
 $ tockloader listen
 Using "/dev/cu.usbserial-c098e513000c - Hail IoT Module - TockOS"

 Listening for serial output.
 ProcessConsole::start

>  Starting process console

Reviewer: I don't think these messages are shown.

 Process Console
 ===============
 

>  `process_console` is a capsule that implements a text console over the UART that allows

Reviewer: ```suggestion
 `process_console` is a capsule that implements a small shell over the UART that allows
```

>  a terminal to inspect and control userspace processes.

Reviewer: ```suggestion
 a terminal to inspect the kernel and control userspace processes.
```

>  Protocol

Reviewer: ```suggestion
Commands
```

 --------

 This module provides a simple text-based console to inspect and control
 which processes are running. The console has eleven commands:
  - `help` - prints the available commands and arguments
  - `status` - prints the current system status
  - `list` - lists the current processes with their IDs and running state
  - `stop n` - stops the process with name n
  - `start n` - starts the stopped process with name n
  - `fault n` - forces the process with name n into a fault state
  - `terminate n` - terminates the running process with name n, moving to the Terminated state

>   - `boot n` - tries to boot a Terminated process with name n

Reviewer: What does `boot` mean? Restart?

Author: It means restart, I've changed the description.

 - `Quanta`: How many times this process has exceeded its allotted time
   quanta.
 - `Syscalls`: The number of system calls the process has made to the kernel.
 - `Restarts`: How many times this process has crashed and been restarted by
   the kernel.
 - `Grants`: The number of grants that have been initialized for the process
   out of the total number of grants defined by the kernel.
 - `State`: The state the process is in.

 Setup
 -----

 You need a device that provides the `hil::uart::UART` trait. This code
 connects a `ProcessConsole` directly up to USART0:

 ```rust
 # use kernel::{capabilities, hil, static_init};
 # use capsules::process_console::ProcessConsole;

 pub struct Capability;
 unsafe impl capabilities::ProcessManagementCapability for Capability {}

 let pconsole = static_init!(
     ProcessConsole<usart::USART>,
     ProcessConsole::new(&usart::USART0,
                  115200,
                  &mut console::WRITE_BUF,
                  &mut console::READ_BUF,
                  &mut console::COMMAND_BUF,
                  kernel,
                  Capability));
 hil::uart::UART::set_client(&usart::USART0, pconsole);

 pconsole.start();
 ```

> 

Reviewer: ```suggestion
> Note: Using the process console might require allocating more stack to the kernel. This is done by modifying ...
```

 Using ProcessConsole
 --------------------

>  With this capsule properly added to a board's `main.rs` and that kernel

Reviewer: What does `that kernel` mean here?

Author: It means the Tock kernel, I've changed the documentation.

 loaded to the board, make sure there is a serial connection to the board.
 Likely, this just means connecting a USB cable from a computer to the board.
 Next, establish a serial console connection to the board. An easy way to do
 this is to run:

 ```shell
 $ tockloader listen
[INFO   ] No device name specified. Using default name "tock".
[INFO   ] No serial port with device name "tock" found.
[INFO   ] Found 2 serial ports.
Multiple serial port options found. Which would you like to use?
[0]	/dev/ttyS4 - n/a
[1]	/dev/ttyACM0 - "BBC micro:bit CMSIS-DAP" - mbed Serial Port

Which option? [0] 1
[INFO   ] Using "/dev/ttyACM0 - "BBC micro:bit CMSIS-DAP" - mbed Serial Port".
[INFO   ] Listening for serial output.

tock$ 
 ```

 For the examples below we will have 2 processes on the board: blink (which will `blink` all the LEDs that are 
 connected to the kernel), and `c_hello` (which prints 'Hello World' when the console is started). Also, a micro:bit v2 board was used as support for the commands, so the results may vary on other devices.

>  

Reviewer: ```suggestion
 We will assume that the user has a serial connection to the board, either by using tockloader or another serial port software.
```
